web/source/keymanweb.js

```diff
@@ -1977,62 +2131,242 @@ if(!window['tavultesoft']['keymanweb']['initialized']) {
+     * Returns     None
+     * Description Legacy:  Enable KMW control element 
+     */    
     keymanweb.EnableControl = function(Pelem)
     {
-      var Ln, Lc;
-      for(Ln=0; Ln<keymanweb._Controls.length; Ln++) // I1511 - array prototype extended
-        if(keymanweb._Controls[Ln].LControl == Pelem)
-        {
-          keymanweb._Controls[Ln].LEnabled = 1;
-          return;
+      if(!keymanweb.isAttached(Pelem) || !Pelem._kmwAttachment.legacy) {
+        var Lc = {LControl:Pelem, LEnabled:1, LDefaultInternalName:'-'};
+        keymanweb._Controls=keymanweb._push(keymanweb._Controls,Lc);
+        if(keymanweb.isAttached(Pelem)) {
+          Pelem._kmwAttachment.legacy = Lc;
         }
-      Lc = {LControl:Pelem, LEnabled:1, LDefaultInternalName:'-'};
-      keymanweb._Controls=keymanweb._push(keymanweb._Controls,Lc);
+      } else {
+        Pelem._kmwAttachment.legacy.LEnabled = 1;
+      }
     }
     
     /**
+     * Function     setKeyboardForControl
+     * Scope        Public   
+     * @param       {Element}    Pelem    Control element 
+     * @param       {string=}    Pkbd     Keyboard (Clears the set keyboard if set to null.)  
+     * @param       {string|null=}     Plc      Language Code
+     * Description  Set default keyboard for the control 
+     */    
+    keymanweb['setKeyboardForControl'] = keymanweb.setKeyboardForControl = function(Pelem, Pkbd, Plc) {
+      /* pass null for kbd to specify no default, or '' to specify English */
+      var index = Pkbd.indexOf("Keyboard_");
+      if(index < 0 && Pkbd != '') {
+        Pkbd = "Keyboard_" + Pkbd;
+      }
+
+      if(Pkbd == null) {
+        Plc = null;
+      }
+
+      if(Pelem.tagName.toLowerCase() == "iframe") {
+        console.warn("'keymanweb.setKeyboardForControl' cannot set keyboard on iframes.");
+        return;
+      }
+
+      if(!keymanweb.isAttached(Pelem)) {
+        console.error("KeymanWeb is not attached to element " + Pelem);
+        return;
+      } else if(Pelem._kmwAttachment.legacy) {
+        console.warn("Correct behavior of KeymanWeb cannot be guaranteed when used alongside certain legacy functions in use.");
+        keymanweb.legacy['SetDefaultKeyboardForControl'](Pelem, Pkbd);
+        return;
+      } else {
+        Pelem._kmwAttachment.keyboard = Pkbd;
+        Pelem._kmwAttachment.languageCode = Plc;
+
+        // If Pelem is the focused element/active control, we should set the keyboard in place now.
+        // 'kmw_ip' is the touch-alias for the original page's control.
+        if((keymanweb._LastActiveElement == Pelem || keymanweb._LastActiveElement == Pelem['kmw_ip']) 
+            && keymanweb._LastActiveElement) {
+            
+          if(Pkbd != null && Plc != null) { // Second part necessary for Closure.
+            keymanweb.setActiveKeyboard(Pkbd, Plc);
+          } else {
+            keymanweb.setActiveKeyboard(keymanweb.globalKeyboard, keymanweb.globalLanguageCode);
```

Comment:
If we are clearing the keyboard property on the element, why do we need to change the active keyboard afterwards?